veles/secrets/gitlabpat/detector.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gitlabpat contains a Veles Secret type and a Detector for
+// Gitlab Personal Access Tokens (prefix `glpat-`).
+package gitlabpat
+
+import (
+	"hash/crc32"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/google/osv-scalibr/veles"
+)
+
+// maxTokenLength is the maximum size of a Gitlab personal access token.
+const maxTokenLength = 319
+
+// Regular expressions for GitLab Personal Access Tokens:
+//
+// Based on the specs at: https://gitlab.com/gitlab-com/content-sites/handbook/-/blob/a5c49599bd88f1751616b40e4e32331aa2c8bf50/content/handbook/engineering/architecture/design-documents/cells/routable_tokens.md#L80
+var (
+	reRoutableVersioned = regexp.MustCompile(`(?P<prefix>glpat-)(?P<payload>[0-9A-Za-z_-]{27,300})\.[0-9a-z]{2}\.(?P<length>[0-9a-z]{2})(?P<crc>[0-9a-z]{7})`)
+	reRoutable          = regexp.MustCompile(`glpat-[0-9A-Za-z_-]{27,300}\.[0-9a-z]{2}[0-9a-z]{7}`)
+	reLegacy            = regexp.MustCompile(`glpat-[0-9A-Za-z_-]{20}`)
+)
+
+var _ veles.Detector = NewDetector()
+
+// isValidCRC32 validates the CRC32 checksum of a GitLab PAT.
+// According to the documentation, the CRC32 is calculated on <prefix><base64-payload>.<base64-payload-length>
```

Comment:
This may not work with versioned routable tokens. I think the doc forgot to update the sections of CRC32 calculation. https://gitlab.com/gitlab-com/content-sites/handbook/-/commit/89f9377913f52106ec75bf5f5f3eefb0b185b36c

https://handbook.gitlab.com/handbook/engineering/architecture/design-documents/cells/routable_tokens/#implementation has a link to the code, and the CRC just used everything before the checksum. https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/authn/token_field/generator/routable_token.rb#L69-83

https://gitlab.com/gitlab-com/content-sites/handbook/-/commit/89f9377913f52106ec75bf5f5f3eefb0b185b36c

Reply:
Thanks. This looks good, and seems working with a PAT I generated. Please update the comment `<prefix><base64-payload>.<base64-payload-length>`-> `<prefix><base64-payload>.<token-version>.<base64-payload-length>`

Also, please also mention in the comment that this function is for versioned routable tokens.